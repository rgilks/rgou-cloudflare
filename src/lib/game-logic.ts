import {
  GameState,
  Player,
  Move,
  PiecePosition,
  ROSETTE_SQUARES,
  TRACK_LENGTH,
  PIECES_PER_PLAYER,
  PLAYER1_TRACK,
  PLAYER2_TRACK,
} from "./types";

// Initialize a new game state
export function initializeGame(): GameState {
  const player1Pieces: PiecePosition[] = Array(PIECES_PER_PLAYER)
    .fill(null)
    .map(() => ({
      square: -1, // Starting position
      player: "player1" as Player,
    }));

  const player2Pieces: PiecePosition[] = Array(PIECES_PER_PLAYER)
    .fill(null)
    .map(() => ({
      square: -1, // Starting position
      player: "player2" as Player,
    }));

  return {
    board: Array(TRACK_LENGTH).fill(null),
    player1Pieces,
    player2Pieces,
    currentPlayer: "player1",
    gameStatus: "playing",
    winner: null,
    diceRoll: null,
    canMove: false,
    validMoves: [],
  };
}

// Roll the ancient dice (4 binary dice, count the number of marked corners)
export function rollDice(): number {
  let count = 0;
  for (let i = 0; i < 4; i++) {
    if (Math.random() < 0.5) count++;
  }
  return count;
}

// Get the track path for a specific player
function getPlayerTrack(player: Player): number[] {
  return player === "player1" ? PLAYER1_TRACK : PLAYER2_TRACK;
}

// Get the actual board position for a player's track position
function getActualPosition(player: Player, trackPosition: number): number {
  const track = getPlayerTrack(player);
  return trackPosition >= 0 && trackPosition < track.length
    ? track[trackPosition]
    : -1;
}

// Check if a square is a rosette (safe square)
function isRosette(square: number): boolean {
  return ROSETTE_SQUARES.includes(square);
}

// Get valid moves for the current player
export function getValidMoves(gameState: GameState): number[] {
  if (!gameState.diceRoll || gameState.diceRoll === 0) return [];

  const currentPieces =
    gameState.currentPlayer === "player1"
      ? gameState.player1Pieces
      : gameState.player2Pieces;

  const validMoves: number[] = [];

  currentPieces.forEach((piece, index) => {
    const currentTrackPos =
      piece.square === -1
        ? -1
        : getPlayerTrack(gameState.currentPlayer).indexOf(piece.square);
    const newTrackPos = currentTrackPos + gameState.diceRoll!;

    // Check if move is valid
    if (newTrackPos >= getPlayerTrack(gameState.currentPlayer).length) {
      // Moving off the board (finishing) - only valid if exact
      if (newTrackPos === getPlayerTrack(gameState.currentPlayer).length) {
        validMoves.push(index);
      }
    } else {
      const newActualPos = getActualPosition(
        gameState.currentPlayer,
        newTrackPos
      );
      const occupant = gameState.board[newActualPos];

      // Can move if square is empty, or occupied by opponent (and not on rosette)
      if (
        !occupant ||
        (occupant.player !== gameState.currentPlayer &&
          !isRosette(newActualPos))
      ) {
        validMoves.push(index);
      }
    }
  });

  return validMoves;
}

// Make a move
export function makeMove(gameState: GameState, pieceIndex: number): GameState {
  if (!gameState.validMoves.includes(pieceIndex) || !gameState.diceRoll) {
    return gameState; // Invalid move
  }

  const newState = { ...gameState };
  const isPlayer1 = gameState.currentPlayer === "player1";
  const currentPieces = isPlayer1
    ? [...gameState.player1Pieces]
    : [...gameState.player2Pieces];
  const piece = currentPieces[pieceIndex];

  // Calculate new position
  const currentTrackPos =
    piece.square === -1
      ? -1
      : getPlayerTrack(gameState.currentPlayer).indexOf(piece.square);
  const newTrackPos = currentTrackPos + gameState.diceRoll;

  // Remove piece from old position
  if (piece.square >= 0) {
    newState.board[piece.square] = null;
  }

  // Check if finishing
  if (newTrackPos >= getPlayerTrack(gameState.currentPlayer).length) {
    currentPieces[pieceIndex] = { ...piece, square: 20 }; // Finished
  } else {
    const newActualPos = getActualPosition(
      gameState.currentPlayer,
      newTrackPos
    );
    const occupant = gameState.board[newActualPos];

    // If there's an opponent piece, send it back to start
    if (occupant && occupant.player !== gameState.currentPlayer) {
      const opponentPieces = isPlayer1
        ? [...gameState.player2Pieces]
        : [...gameState.player1Pieces];
      const opponentPieceIndex = opponentPieces.findIndex(
        (p) => p.square === newActualPos
      );
      if (opponentPieceIndex >= 0) {
        opponentPieces[opponentPieceIndex] = {
          ...opponentPieces[opponentPieceIndex],
          square: -1,
        };
        if (isPlayer1) {
          newState.player2Pieces = opponentPieces;
        } else {
          newState.player1Pieces = opponentPieces;
        }
      }
    }

    // Place piece in new position
    currentPieces[pieceIndex] = { ...piece, square: newActualPos };
    newState.board[newActualPos] = currentPieces[pieceIndex];
  }

  // Update pieces
  if (isPlayer1) {
    newState.player1Pieces = currentPieces;
  } else {
    newState.player2Pieces = currentPieces;
  }

  // Check for win condition
  const finishedPieces = currentPieces.filter((p) => p.square === 20).length;
  if (finishedPieces === PIECES_PER_PLAYER) {
    newState.gameStatus = "finished";
    newState.winner = gameState.currentPlayer;
  }

  // Determine next player (stay if landed on rosette or if game finished)
  const landedOnRosette =
    newTrackPos < getPlayerTrack(gameState.currentPlayer).length &&
    isRosette(getActualPosition(gameState.currentPlayer, newTrackPos));

  if (!landedOnRosette && newState.gameStatus !== "finished") {
    newState.currentPlayer =
      gameState.currentPlayer === "player1" ? "player2" : "player1";
  }

  // Reset move state
  newState.diceRoll = null;
  newState.canMove = false;
  newState.validMoves = [];

  return newState;
}

// Process dice roll
export function processDiceRoll(gameState: GameState): GameState {
  const diceRoll = rollDice();
  const newState = {
    ...gameState,
    diceRoll,
    canMove: diceRoll > 0,
    validMoves: diceRoll > 0 ? getValidMoves({ ...gameState, diceRoll }) : [],
  };

  // If no valid moves and dice roll > 0, skip turn
  if (diceRoll > 0 && newState.validMoves.length === 0) {
    return {
      ...newState,
      currentPlayer:
        gameState.currentPlayer === "player1" ? "player2" : "player1",
      diceRoll: null,
      canMove: false,
      validMoves: [],
    };
  }

  return newState;
}
