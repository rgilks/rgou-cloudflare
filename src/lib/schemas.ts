import { z } from 'zod';

export const PlayerSchema = z.enum(['player1', 'player2']);
export type Player = z.infer<typeof PlayerSchema>;

export const MoveTypeSchema = z.enum(['move', 'capture', 'rosette', 'finish']);
export type MoveType = z.infer<typeof MoveTypeSchema>;

export const GameStatusSchema = z.enum(['waiting', 'playing', 'finished']);
export type GameStatus = z.infer<typeof GameStatusSchema>;

export const PiecePositionSchema = z
  .object({
    square: z.number(),
    player: PlayerSchema,
  })
  .refine(val => val.square === -1 || val.square === 20 || (val.square >= 0 && val.square < 21), {
    message: 'square must be -1 (start), 0-20 (board), or 20 (finished)',
    path: ['square'],
  });
export type PiecePosition = z.infer<typeof PiecePositionSchema>;

export const MoveRecordSchema = z.object({
  player: PlayerSchema,
  diceRoll: z.number(),
  pieceIndex: z.number(),
  fromSquare: z.number(),
  toSquare: z.number(),
  moveType: MoveTypeSchema.nullable(),
});
export type MoveRecord = z.infer<typeof MoveRecordSchema>;

export const GameStateSchema = z.object({
  board: z.array(PiecePositionSchema.nullable()).length(21),
  player1Pieces: z.array(PiecePositionSchema),
  player2Pieces: z.array(PiecePositionSchema),
  currentPlayer: PlayerSchema,
  gameStatus: GameStatusSchema,
  winner: PlayerSchema.nullable(),
  diceRoll: z.number().nullable(),
  canMove: z.boolean(),
  validMoves: z.array(z.number()),
  history: z.array(MoveRecordSchema),
  startTime: z.number().optional(),
});
export type GameState = z.infer<typeof GameStateSchema>;

export const MoveSchema = z.object({
  pieceIndex: z.number(),
  diceRoll: z.number(),
  player: PlayerSchema,
  newSquare: z.number(),
  moveType: MoveTypeSchema.nullable(),
});
export type Move = z.infer<typeof MoveSchema>;

export const GameStatsSchema = z.object({
  wins: z.number(),
  losses: z.number(),
  gamesPlayed: z.number(),
});
export type GameStats = z.infer<typeof GameStatsSchema>;

export const GameModeSchema = z.enum(['play', 'watch']);
export type GameMode = z.infer<typeof GameModeSchema>;

export const GameActionSchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('ROLL_DICE') }),
  z.object({ type: z.literal('MAKE_MOVE'), move: MoveSchema }),
  z.object({ type: z.literal('RESET_GAME') }),
  z.object({ type: z.literal('AI_MOVE'), move: MoveSchema }),
]);
export type GameAction = z.infer<typeof GameActionSchema>;

export const MoveEvaluationSchema = z.object({
  pieceIndex: z.number(),
  score: z.number(),
  moveType: z.string(),
  fromSquare: z.number(),
  toSquare: z.number().nullable(),
});
export type MoveEvaluation = z.infer<typeof MoveEvaluationSchema>;

export const DiagnosticsSchema = z.object({
  searchDepth: z.number(),
  validMoves: z.array(z.number()),
  moveEvaluations: z.array(MoveEvaluationSchema),
  transpositionHits: z.number(),
  nodesEvaluated: z.number(),
});
export type Diagnostics = z.infer<typeof DiagnosticsSchema>;

export const TimingsSchema = z.object({
  aiMoveCalculation: z.number(),
  totalHandlerTime: z.number(),
});
export type Timings = z.infer<typeof TimingsSchema>;

export const AIResponseSchema = z.object({
  move: z.number().nullable(),
  evaluation: z.number(),
  thinking: z.string(),
  timings: TimingsSchema,
  diagnostics: DiagnosticsSchema,
  aiType: z.enum(['client', 'server', 'fallback', 'ml']),
});
export type AIResponse = z.infer<typeof AIResponseSchema>;

export const ServerAIResponseSchema = AIResponseSchema.omit({ aiType: true });
export type ServerAIResponse = z.infer<typeof ServerAIResponseSchema>;

export const SaveGamePayloadSchema = z.object({
  winner: PlayerSchema,
  history: z.array(MoveRecordSchema),
  playerId: z.string(),
  moveCount: z.number().optional(),
  duration: z.number().optional(),
  clientHeader: z.string().optional(),
  gameType: z.string().default('standard'),
  ai1Version: z.string().optional(),
  ai2Version: z.string().optional(),
  gameVersion: z.string().optional(),
});
export type SaveGamePayload = z.infer<typeof SaveGamePayloadSchema>;

export const GameConstants = {
  ROSETTE_SQUARES: [0, 7, 13, 15, 16] as const,
  TRACK_LENGTH: 20,
  BOARD_ARRAY_SIZE: 21,
  PIECES_PER_PLAYER: 7,
  PLAYER1_TRACK: [3, 2, 1, 0, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] as const,
  PLAYER2_TRACK: [19, 18, 17, 16, 4, 5, 6, 7, 8, 9, 10, 11, 14, 15] as const,
} as const;
