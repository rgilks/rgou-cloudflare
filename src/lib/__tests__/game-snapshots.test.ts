import { describe, it, expect } from 'vitest';
import { initializeGame, makeMove } from '../game-logic';
import { GameConstants } from '../schemas';

const LAST_PLAYER1_SQUARE = GameConstants.PLAYER1_TRACK[GameConstants.PLAYER1_TRACK.length - 1];

// For capture test
const CAPTURE_FROM = GameConstants.PLAYER1_TRACK[1]; // 2
const CAPTURE_TO = GameConstants.PLAYER1_TRACK[2]; // 1

describe('Game State Snapshots', () => {
  it('should maintain consistent state after capture', () => {
    const game = initializeGame();
    game.player1Pieces[0].square = CAPTURE_FROM;
    game.board[CAPTURE_FROM] = game.player1Pieces[0];
    game.player2Pieces[0].square = CAPTURE_TO;
    game.board[CAPTURE_TO] = game.player2Pieces[0];
    game.diceRoll = 1;
    game.canMove = true;
    game.validMoves = [0];

    const [newState] = makeMove(game, 0);

    expect(newState.player1Pieces[0].square).toBe(CAPTURE_TO);
    expect(newState.board[CAPTURE_TO]).toEqual(newState.player1Pieces[0]);
    expect(newState.player2Pieces[0].square).toBe(-1);
  });

  it('should maintain consistent state after rosette landing', () => {
    const game = initializeGame();
    game.diceRoll = 4;
    game.canMove = true;
    game.validMoves = [0];

    const [newState] = makeMove(game, 0);

    expect(newState.board[0]).toEqual(newState.player1Pieces[0]);
    expect(newState.currentPlayer).toBe('player1');
  });

  it('should maintain consistent state after piece finish', () => {
    const game = initializeGame();
    game.player1Pieces[0].square = LAST_PLAYER1_SQUARE;
    game.board[LAST_PLAYER1_SQUARE] = game.player1Pieces[0];
    for (let i = 1; i < 7; i++) {
      game.player1Pieces[i].square = 20;
    }
    game.diceRoll = 1;
    game.canMove = true;
    game.validMoves = [0];

    const [newState] = makeMove(game, 0);

    expect(newState.player1Pieces[0].square).toBe(20);
    expect(newState.board[LAST_PLAYER1_SQUARE]).toBeNull();
    expect(newState.gameStatus).toBe('finished');
    expect(newState.winner).toBe('player1');
  });

  it('should maintain consistent state after game win', () => {
    const game = initializeGame();
    for (let i = 0; i < 6; i++) {
      game.player1Pieces[i].square = 20;
    }
    game.player1Pieces[6].square = LAST_PLAYER1_SQUARE;
    game.board[LAST_PLAYER1_SQUARE] = game.player1Pieces[6];
    game.diceRoll = 1;
    game.canMove = true;
    game.validMoves = [6];

    const [newState] = makeMove(game, 6);

    expect(newState.player1Pieces[6].square).toBe(20);
    expect(newState.board[LAST_PLAYER1_SQUARE]).toBeNull();
    expect(newState.gameStatus).toBe('finished');
    expect(newState.winner).toBe('player1');
  });

  it('should maintain consistent state after multiple moves', () => {
    const game = initializeGame();
    const moves = [
      { pieceIndex: 0, diceRoll: 4 },
      { pieceIndex: 1, diceRoll: 3 },
      { pieceIndex: 2, diceRoll: 2 },
    ];

    let currentState = game;
    for (const move of moves) {
      currentState.diceRoll = move.diceRoll;
      currentState.canMove = true;
      currentState.validMoves = [move.pieceIndex];
      [currentState] = makeMove(currentState, move.pieceIndex);
    }

    expect(currentState.player1Pieces[0].square).toBe(0);
    expect(currentState.player1Pieces[1].square).toBe(1);
    expect(currentState.player1Pieces[2].square).toBe(-1);
  });

  it('should handle edge case: all pieces finished', () => {
    const game = initializeGame();
    for (let i = 0; i < 7; i++) {
      game.player1Pieces[i].square = 20;
    }
    game.diceRoll = 1;
    game.canMove = false;
    game.validMoves = [];
    game.gameStatus = 'finished';
    game.winner = 'player1';

    expect(game.gameStatus).toBe('finished');
    expect(game.winner).toBe('player1');
  });
});
