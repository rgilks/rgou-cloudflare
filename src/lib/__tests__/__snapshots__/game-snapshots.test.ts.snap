// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game State Snapshots > After First Move > should match state after player 1 first move > after-player1-first-move 1`] = `
{
  "board": [
    {
      "player": "player1",
      "square": 0,
    },
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player1",
  "diceRoll": null,
  "gameStatus": "playing",
  "history": [
    {
      "diceRoll": 4,
      "fromSquare": -1,
      "moveType": "rosette",
      "pieceIndex": 0,
      "player": "player1",
      "toSquare": 0,
    },
  ],
  "player1Pieces": [
    {
      "player": "player1",
      "square": 0,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": null,
}
`;

exports[`Game State Snapshots > Capture Scenario > should match state after capture > after-capture-move 1`] = `
{
  "board": [
    null,
    null,
    null,
    null,
    null,
    null,
    {
      "player": "player1",
      "square": 6,
    },
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player2",
  "diceRoll": null,
  "gameStatus": "playing",
  "history": [
    {
      "diceRoll": 2,
      "fromSquare": 4,
      "moveType": "capture",
      "pieceIndex": 0,
      "player": "player1",
      "toSquare": 6,
    },
  ],
  "player1Pieces": [
    {
      "player": "player1",
      "square": 6,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": null,
}
`;

exports[`Game State Snapshots > Finish Move > should match state after finish move > after-finish-move 1`] = `
{
  "board": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player2",
  "diceRoll": null,
  "gameStatus": "playing",
  "history": [
    {
      "diceRoll": 1,
      "fromSquare": 13,
      "moveType": "finish",
      "pieceIndex": 0,
      "player": "player1",
      "toSquare": 20,
    },
  ],
  "player1Pieces": [
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": null,
}
`;

exports[`Game State Snapshots > Game Win Scenario > should match state after game win > after-game-win 1`] = `
{
  "board": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player1",
  "diceRoll": null,
  "gameStatus": "finished",
  "history": [
    {
      "diceRoll": 1,
      "fromSquare": 13,
      "moveType": "finish",
      "pieceIndex": 6,
      "player": "player1",
      "toSquare": 20,
    },
  ],
  "player1Pieces": [
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": "player1",
}
`;

exports[`Game State Snapshots > Initial Game State > should match initial game state snapshot > initial-game-state 1`] = `
{
  "board": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player1",
  "diceRoll": null,
  "gameStatus": "playing",
  "history": [],
  "player1Pieces": [
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": null,
}
`;

exports[`Game State Snapshots > No Valid Moves > should match state when no valid moves available > no-valid-moves 1`] = `
{
  "board": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player2",
  "diceRoll": null,
  "gameStatus": "playing",
  "history": [],
  "player1Pieces": [
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
    {
      "player": "player1",
      "square": 20,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": null,
}
`;

exports[`Game State Snapshots > Roll of Zero > should match state after roll of zero > after-roll-of-zero 1`] = `
{
  "board": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player2",
  "diceRoll": null,
  "gameStatus": "playing",
  "history": [],
  "player1Pieces": [
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": null,
}
`;

exports[`Game State Snapshots > Rosette Scenario > should match state after rosette move > after-rosette-move 1`] = `
{
  "board": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    {
      "player": "player1",
      "square": 7,
    },
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "canMove": false,
  "currentPlayer": "player1",
  "diceRoll": null,
  "gameStatus": "playing",
  "history": [
    {
      "diceRoll": 2,
      "fromSquare": 5,
      "moveType": "rosette",
      "pieceIndex": 0,
      "player": "player1",
      "toSquare": 7,
    },
  ],
  "player1Pieces": [
    {
      "player": "player1",
      "square": 7,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
    {
      "player": "player1",
      "square": -1,
    },
  ],
  "player2Pieces": [
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
    {
      "player": "player2",
      "square": -1,
    },
  ],
  "validMoves": [],
  "winner": null,
}
`;
